examples/instance_refresh/main.tf

```diff
@@ -173,29 +192,35 @@ resource "helm_release" "anth" {
   }
   set {
     name  = "serviceAccount.annotations.eks\\.amazonaws\\.com/role-arn"
-    value = module.node_term_role.iam_role_arn
+    value = module.aws_node_termination_handler_role.iam_role_arn
     type  = "string"
   }
   set {
     name  = "enableSqsTerminationDraining"
     value = "true"
   }
+  set {
+    name  = "enableSpotInterruptionDraining"
+    value = "true"
+  }
   set {
     name  = "queueURL"
-    value = module.node_term_sqs.sqs_queue_id
+    value = module.aws_node_termination_handler_sqs.sqs_queue_id
   }
   set {
     name  = "logLevel"
-    value = "DEBUG"
+    value = "debug"
   }
 }
 
 # Creating the lifecycle-hook outside of the ASG resource's `initial_lifecycle_hook`
 # ensures that node termination does not require the lifecycle action to be completed,
 # and thus allows the ASG to be destroyed cleanly.
-resource "aws_autoscaling_lifecycle_hook" "node_term" {
-  name                   = "node_term-${local.cluster_name}"
-  autoscaling_group_name = module.eks.workers_asg_names[0]
+resource "aws_autoscaling_lifecycle_hook" "aws_node_termination_handler" {
+  for_each = toset(module.eks.workers_asg_names)
```

Comment:
I am not sure, but maybe adding the termination handler to each ASG might cause some delays in node draining if the ASG is not properly tagged for ANTH.

Reply:
This is because for_each needs to know keys during the refresh staage. That means, we can't use for_each here, but `count` will do the trick.

Reply:
The following seems to resolve the issue:

I can commit it under #1370 

```
diff --git a/examples/instance_refresh/main.tf b/examples/instance_refresh/main.tf
index 7c60621..1b87d8e 100644
--- a/examples/instance_refresh/main.tf
+++ b/examples/instance_refresh/main.tf
@@ -217,10 +217,9 @@ resource "helm_release" "aws_node_termination_handler" {
 # ensures that node termination does not require the lifecycle action to be completed,
 # and thus allows the ASG to be destroyed cleanly.
 resource "aws_autoscaling_lifecycle_hook" "aws_node_termination_handler" {
-  for_each = toset(module.eks.workers_asg_names)
-
+  count                  = length(module.eks.workers_asg_names)
   name                   = "aws-node-termination-handler"
-  autoscaling_group_name = each.value
+  autoscaling_group_name = module.eks.workers_asg_names[count.index]
   lifecycle_transition   = "autoscaling:EC2_INSTANCE_TERMINATING"
   heartbeat_timeout      = 300
   default_result         = "CONTINUE"
```

Reply:
Sorry guys, our implementation is a module wrapping the `terraform-aws-eks` module and containing sub modules for doing things like installing the `aws-node-termination-handler` and I copied our config out and flattened it.

Reply:
Hello, I'm trying to get the asg name and arns output of my node_group with `module.eks.workers_asg_names` but the output is empty.

Am I missing something when using node_groups (not self-managed nodes) ?

Reply:
`module.eks.workers_asg_names` is for self-managed node group only. If you're using managed node group,  there is `node_groups` output.

I don't think you need node termination handler if you're using managed node group. AWS handle node draining for you. Please see AWS docs for managed node groups.

Reply:
@barryib even when using spot instances?

Reply:
@andronux spot termination for managed node groups should be automated.

Reply:
ok thank you @stevehipwell @barryib !

Reply:
To me, you don't need it for MNG. AWS handle drains for you. Please see:
- https://aws.amazon.com/fr/about-aws/whats-new/2020/12/amazon-eks-support-ec2-spot-instances-managed-node-groups/
- https://aws.amazon.com/fr/blogs/containers/amazon-eks-now-supports-provisioning-and-managing-ec2-spot-instances-in-managed-node-groups/ (spot section)
- https://docs.aws.amazon.com/eks/latest/userguide/managed-node-groups.html (spot section)

Or I'm probably missing something.